### PR TITLE
feat(lbf): generalise wrapper and logger

### PR DESCRIPTION
**Pull request by WiemKhlifi**
## What?
Based on the changes made in this **first PR** #948 for the **RwareWrapper**, we added a separate **LbfWrapper** as well as some changes to support the LBF environment. -> Merged them into one single wrapper 🤞 
## Why?
LbfWrapper is useful in this case to choose what reward will be used during training whether it is the shared reward or not.
## How?
* Create `LbfWrapper` and separate file systems for LBF -> changed this one general wrapper
* Edit the `logger_utils` file to log LBF tasks.
* Create a folder for different config files for LBF.

### CODE REVIEW

File: mava/configs/env/lbf.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+defaults:
+  - _self_
+  - lbf_scenario: Foraging-2s-8x8-2p-2f-coop # [Foraging-2s-8x8-2p-2f-coop, Foraging-8x8-2p-2f-coop, Foraging-2s-10x10-3p-3f, Foraging-10x10-3p-3f, Foraging-15x15-4p-3f, Foraging-15x15-4p-5f]
+
+env_name: LevelBasedForaging-v0
+task_name: Foraging-2s-8x8-2p-2f-coop
+# Choose whether to use the use_idividual_reward = True OR aggregate_rewards=True.
+use_idividual_reward: False # If True, the network uses the list of different rewards given.
+aggregate_rewards : True  # If True, aggregates rewards across agents using the specified function.
+aggregate_rewards_with: sum # The function for aggregating rewards ("sum" or "mean").
```
Comment (RuanJohn) [inline mava/configs/env/lbf.yaml:11]: NIT: I think we could name this `reward_aggregation_function` ?

File: mava/configs/env/lbf_scenario/Foraging-15x15-4p-3f.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# The config of the Foraging-15x15-4p-3f scenario
+
+task_config_gen:
+  grid_size: 15
+  fov : 15
```
Comment (RuanJohn) [inline mava/configs/env/lbf_scenario/Foraging-15x15-4p-3f.yaml:5]: A general description somewhere for what these attributes are would be really nice

File: mava/configs/env/lbf_scenario/Foraging-10x10-3p-3f.yaml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# The config of the Foraging-10x10-3p-3f scenario
+
+task_config:
```
Comment (RuanJohn) [inline mava/configs/env/lbf_scenario/Foraging-10x10-3p-3f.yaml:3]: A general comment on the config naming. I think it will be nicer to to remove `Foraging-` from all the LBF scenario names since all the scenario configs are already in the `lbf_scenario` folder.

File: mava/systems/ff_ippo_lbf.py [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+        elapsed_time = time.time() - start_time
+        evaluator_output.episodes_info["steps_per_second"] = steps_per_rollout / elapsed_time
+        log(
+            metrics=evaluator_output,
+            t_env=steps_per_rollout * (i + 1),
+            absolute_metric=True,
+        )
+
+
+@hydra.main(config_path="../configs", config_name="default_ff_ippo.yaml", version_base="1.2")
```
Comment (RuanJohn) [inline mava/systems/ff_ippo_lbf.py:625]: It will also be good if we have a default_ff_ippo_lbf config or something along those lines? Because this will probably not do what we want if the default ppo config points to rware.

File: mava/wrappers/jumanji.py [hunk 5]
```diff
@@ -93,6 +59,84 @@ def step(
         return state, timestep
 
 
+class MultiAgentWrapper(Wrapper):
```
Comment (RuanJohn) [inline mava/wrappers/jumanji.py:62]: I really like this wrapper!!

File: mava/wrappers/jumanji.py [hunk 6]
```diff
@@ -93,6 +59,84 @@ def step(
...
+        self._num_agents = self._env.num_agents
+        # Mapping aggregation functions
+        self.aggregation_functions = {
+            "sum": jnp.sum,
+            "mean": jnp.mean,
+            # Add more functions as needed
+        }
+        self.aggregate_function = self.aggregation_functions.get(
+            aggregate_rewards_with, jnp.sum
+        )  # No error!
```
Comment (RuanJohn) [inline mava/wrappers/jumanji.py:93]: Not sure what this comment means?

File: mava/wrappers/jumanji.py [hunk 7]
```diff
@@ -93,6 +59,84 @@ def step(
...
+
+    def modify_timestep(self, timestep: TimeStep) -> TimeStep[Observation]:
+        """Modify the timestep based on the specified reward handling strategy."""
+        observation = Observation(
+            agents_view=timestep.observation.agents_view,
+            action_mask=timestep.observation.action_mask,
+            step_count=jnp.repeat(timestep.observation.step_count, self._num_agents),
+        )
+
+        if self._use_individual_reward:
```
Comment (RuanJohn) [inline mava/wrappers/jumanji.py:110]: Maybe just a comment at the top explaining that this will replace the timesteps accordingly? It is a bit hard to follow with all the returns.

**Review by RuanJohn — CHANGES_REQUESTED** [review]
Super nice work! :fire: I really like the general env wrapper! Just some changes here and there.

File: mava/configs/env/lbf.yaml (hunk 1 above)
Comment (WiemKhlifi) [reply mava/configs/env/lbf.yaml:11]: Agree!

File: mava/systems/ff_ippo_lbf.py (hunk 4 above)
Comment (WiemKhlifi) [reply mava/systems/ff_ippo_lbf.py:625]: I thought that we simply make changes in the default files but if this makes it easier to run experiments later I'll be happy to quickly add them!

File: mava/configs/env/lbf_scenario/Foraging-10x10-3p-3f.yaml (hunk 3 above)
Comment (WiemKhlifi) [reply mava/configs/env/lbf_scenario/Foraging-10x10-3p-3f.yaml:3]: Sure I'll change them it makes the name longer and it's redundant!

File: mava/configs/env/lbf_scenario/Foraging-15x15-4p-3f.yaml (hunk 2 above)
Comment (WiemKhlifi) [reply mava/configs/env/lbf_scenario/Foraging-15x15-4p-3f.yaml:5]: Sure! Also, should add the config and description for the case if we want to use the GridObserver for LBF? (in this case, the observation is a vector and VectorObserver is set as default).

File: mava/wrappers/jumanji.py (hunk 6 above)
Comment (WiemKhlifi) [reply mava/wrappers/jumanji.py]: Sorry it's a typo.

File: mava/wrappers/jumanji.py (hunk 7 above)
Comment (WiemKhlifi) [reply mava/wrappers/jumanji.py:120]: I added some details as comments and docs, should we add more or change something?

File: mava/wrappers/jumanji.py (hunk 5 above)
Comment (sash-a) [reply mava/wrappers/jumanji.py:27]: Unfortunately this is not going to continue to work. We *need* a wrapper per jumanji env so that we can correctly pack their obs into our `Observation`. 

For example connectors observation looks like this:
```python
class Observation
    grid
    action_mask
    step_count
```

Then many other jumanji envs have different values here. Only lbf and rware have the same obs names that is why this works now, but it won't continue to work, so we need to separate this per env.

File: mava/wrappers/jumanji.py [hunk 8]
```diff
@@ -12,35 +12,87 @@
...
+        aggregate_rewards (bool): If True, aggregates rewards across agents.
+        reward_aggregation_function (str): The function for aggregating rewards ("sum" or "mean").
+    """
+
+    def __init__(
+        self,
+        env: Environment,
+        use_individual_reward: bool = False,
+        aggregate_rewards: bool = False,
+        reward_aggregation_function: str = "sum",
```
Comment (sash-a) [inline mava/wrappers/jumanji.py:43]: I think we shouldn't have these options, we should decide if we want individual/aggregated rewards and use the option we want in each separate wrapper

File: mava/systems/ff_ippo_lbf.py
Comment (sash-a) [inline mava/systems/ff_ippo_lbf.py:1]: Separate files for each env isn't ideal as then our number of files will explode. We want lbf/rware/future envs to share the same files. So we need to make a function e.g `make` that parses our env config and makes the correct env and then we can use that function in all the files.

File: mava/wrappers/jumanji.py (hunk 5 above)
Comment (RuanJohn) [reply mava/wrappers/jumanji.py:27]: I back the idea of a wrapper per env in that case.

File: mava/wrappers/jumanji.py (hunk 8 above)
Comment (RuanJohn) [reply mava/wrappers/jumanji.py:43]: If we are doing a different wrapper per env then hard setting the aggregation is fine.

File: mava/wrappers/jumanji.py [hunk 9]
```diff
@@ -62,3 +55,87 @@ def observation_spec(self) -> specs.Spec[Observation]:
...
+    ):
+        super().__init__(env)
+        self._env: LevelBasedForaging
+        self._aggregate_rewards = aggregate_rewards
+        self.reward_aggregation_function = reward_aggregation_function
+        self._num_agents = self._env.num_agents
+        # Mapping aggregation functions
+        self.aggregation_functions: Dict[str, Callable] = {
+            "sum": jnp.sum,
```
Comment (RuanJohn) [inline mava/wrappers/jumanji.py:102]: I agree with @sash-a's previous comment on this. I think we should just do what the benchmarking paper does and keep that as our default aggregation function. The only boolean we would then need is whether or not to aggregate the rewards over agents or not.

File: mava/configs/system/ff_ippo.yaml [hunk 10]
```diff
@@ -1,6 +1,6 @@
 # --- Defaults FF-IPPO ---
 
-num_updates: 1000 # Number of updates
+num_updates: 5000 # Number of updates
```
Comment (RuanJohn) [inline mava/configs/system/ff_ippo.yaml:3]: Please leave this to 1000 :)

File: mava/configs/env/rware.yaml [hunk 11]
```diff
@@ -4,3 +4,4 @@ defaults:
   - rware_scenario: tiny-2ag # [tiny-2ag, tiny-4ag, tiny-4ag-easy, small-4ag]
 
 env_name: RobotWarehouse-v0
+task_name: tiny-2ag
```
Comment (RuanJohn) [inline mava/configs/env/rware.yaml:7]: I think it is better to leave this in the child configs, otherwise we always have to set the scenario name and the task_name which could lead to accidental mismatches.

File: mava/wrappers/jumanji.py (hunk 8 above)
Comment (WiemKhlifi) [reply mava/wrappers/jumanji.py]: For the aggregation  I let the user choose whether to use the list of individual rewards (default setting) or  set the `aggregate_rewards=True` in lbf config file.

File: mava/configs/env/lbf.yaml [hunk 12]
```diff
@@ -0,0 +1,11 @@
+# ---Environment Configs---
+defaults:
+  - _self_
+  - lbf_scenario: 2s-8x8-2p-2f-coop # [2s-8x8-2p-2f-coop, 8x8-2p-2f-coop, 2s-10x10-3p-3f, 10x10-3p-3f, 15x15-4p-3f, 15x15-4p-5f]
+# Further enviroment config details in "2s-8x8-2p-2f-coop" file.
+
+env_name: LevelBasedForaging-v0
+task_name: 2s-8x8-2p-2f-coop
```
Comment (RuanJohn) [inline mava/configs/env/lbf.yaml:8]: Similarly here, I think it is better to set the task names in the child configs to avoid logging mismatches.

File: mava/wrappers/jumanji.py (hunk 8 above)
Comment (RuanJohn) [reply mava/wrappers/jumanji.py]: This is good, but I don't think we should give an option for _what_ aggregation to use. We should just hardcode it in the wrapper to be the same as what other benchmarks do for each environment.

File: mava/wrappers/jumanji.py (hunk 9 above)
Comment (WiemKhlifi) [reply mava/wrappers/jumanji.py:102]: I'll change the aggregation to true the default  to use the sum of the individual rewards as the benchmarking paper did and make the other case optional (remove the mean function as well).

File: mava/wrappers/jumanji.py [hunk 13]
```diff
@@ -62,3 +55,87 @@ def observation_spec(self) -> specs.Spec[Observation]:
...
+            "mean": jnp.mean,
+            # Add more functions as needed
+        }
+        self.aggregate_function: Callable = self.aggregation_functions.get(
+            reward_aggregation_function, jnp.sum
+        )
```
Comment (sash-a) [inline mava/wrappers/jumanji.py:108]: Bit cleaner

```suggestion

        reward_aggregation_fn: Optional[Callable[[Array], Array] = jnp.sum,
    ):
        super().__init__(env)
        self._env: LevelBasedForaging
        self._num_agents = self._env.num_agents
        self._aggregate_rewards = reward_aggregation_fn is not None
        self.aggregate_function = reward_aggregation_fn
```

**Review by sash-a — COMMENTED** [review]
Slightly cleaner way to do the reward aggregation :+1:

File: mava/wrappers/jumanji.py (hunk 13 above)
Comment (WiemKhlifi) [reply mava/wrappers/jumanji.py]: I set the `jnp.sum` as the only function we'll use as @RuanJohn mentioned, I hope the current changes are better now 🙏

Comment (sash-a) [reply mava/wrappers/jumanji.py]: even better!

File: mava/configs/env/scenario/2s-8x8-2p-2f-coop.yaml [hunk 14]
```diff
@@ -0,0 +1,15 @@
+# The config of the 2s-8x8-2p-2f-coop scenario with the VectorObserver set as default.
+task_name: 2s-8x8-2p-2f-coop
+
+task_config:
+  grid_size: 8 # size of the grid to generate.
+  fov: 2       # field of view of an agent.
+  num_agents: 2 # number of agents on the grid.
+  num_food: 2   # number of food in the environment.
+  max_agent_level: 2 # maximum level of the agents (inclusive).
+  force_coop: True   # force cooperation between agents.
```
Comment (RuanJohn) [inline mava/configs/env/scenario/2s-8x8-2p-2f-coop.yaml:10]: Bit cleaner
```suggestion
  grid_size: 8 # size of the grid to generate.
  fov: 2 # field of view of an agent.
  num_agents: 2 # number of agents on the grid.
  num_food: 2 # number of food in the environment.
  max_agent_level: 2 # maximum level of the agents (inclusive).
  force_coop: True # force cooperation between agents.
```

**Review by RuanJohn — COMMENTED** [review]
One final detail here is that we need to update the requirements to point to a new branch we want to use as the base Jumanji branch so that we have LBF included. 
Do you have a suggestion for where this branch should live @sash-a?

File: mava/wrappers/jumanji.py [hunk 15]
```diff
@@ -17,48 +17,107 @@
...
-        )
-        reward = jnp.repeat(timestep.reward, n_agents)
-        discount = jnp.repeat(timestep.discount, n_agents)
-        return timestep.replace(observation=observation, reward=reward, discount=discount)
+        """Modify the timestep for the selected enviroment."""
+        pass
 
     def reset(self, key: chex.PRNGKey) -> Tuple[State, TimeStep]:
-        """Reset the environment. Updates the step count."""
+        """Reset the environment and updates the timestep."""
```
Comment (sash-a) [inline mava/wrappers/jumanji.py:40]:
```suggestion
        """Reset the environment."""
```

File: mava/wrappers/jumanji.py [hunk 16]
```diff
@@ -17,48 +17,107 @@
...
         state, timestep = self._env.reset(key)
         return state, self.modify_timestep(timestep)
 
     def step(self, state: State, action: chex.Array) -> Tuple[State, TimeStep]:
-        """Step the environment. Updates the step count."""
+        """Step the environment and updates the timestep."""
```
Comment (sash-a) [inline mava/wrappers/jumanji.py:45]:
```suggestion
        """Step the environment."""
```

File: mava/wrappers/jumanji.py [hunk 17]
```diff
@@ -17,48 +17,107 @@
...
-            [0] * self._env.num_agents,
-            [self._env.time_limit] * self._env.num_agents,
+            [0] * self._num_agents,
+            [self.time_limit] * self._num_agents,
             "step_count",
         )
         return self._env.observation_spec().replace(step_count=step_count)
+
+
+class RWAREWrapper(MultiAgentWrapper):
```
Comment (sash-a) [inline mava/wrappers/jumanji.py:61]: I just don't like how it looks with all caps, but I don't mind this too much
```suggestion
class RwareWrapper(MultiAgentWrapper):
```

File: mava/wrappers/jumanji.py [hunk 18]
```diff
@@ -17,48 +17,107 @@
...
+            agents_view=timestep.observation.agents_view,
+            action_mask=timestep.observation.action_mask,
+            step_count=jnp.repeat(timestep.observation.step_count, self._num_agents),
+        )
+        reward = jnp.repeat(timestep.reward, self._num_agents)
+        discount = jnp.repeat(timestep.discount, self._num_agents)
+        return timestep.replace(observation=observation, reward=reward, discount=discount)
+
+
+class LBFWrapper(MultiAgentWrapper):
```
Comment (sash-a) [inline mava/wrappers/jumanji.py:79]: Same here

```suggestion
class LbfWrapper(MultiAgentWrapper):
```

File: mava/wrappers/jumanji.py [hunk 19]
```diff
@@ -17,48 +17,107 @@
...
+    """
+     Multi-agent wrapper for the Level-Based Foraging environment.
+
+    Args:
+        env (Environment): The base environment.
+        aggregate_rewards (bool): If True, aggregates rewards across agents.
+        reward_aggregation_function (str): The function for aggregating rewards ("sum" or "mean").
```
Comment (sash-a) [inline mava/wrappers/jumanji.py:86]:
```suggestion
        use_individual_rewards (bool): If true each agent gets a separate reward, sum reward otherwise.
```

File: mava/wrappers/jumanji.py [hunk 20]
```diff
@@ -17,48 +17,107 @@
...
+    """
+
+    def __init__(
+        self,
+        env: LevelBasedForaging,
+        use_individual_rewards: bool = False,
+    ):
```
Comment (sash-a) [inline mava/wrappers/jumanji.py:93]: Nit: can this fit on a single line?

File: mava/wrappers/jumanji.py [hunk 21]
```diff
@@ -17,48 +17,107 @@
...
 from jumanji.wrappers import Wrapper
 
 from mava.types import Observation, State
 
 
-class RwareWrapper(Wrapper):
-    """Multi-agent wrapper for the Robotic Warehouse environment."""
-
-    def __init__(self, env: RobotWarehouse):
+class MultiAgentWrapper(Wrapper):
```
Comment (sash-a) [inline mava/wrappers/jumanji.py:29]: This is a great base class for our future environments :fire:

**Review by sash-a — CHANGES_REQUESTED** [review]
Very minor nitpicks, otherwise I'm happy :rocket:

File: mava/wrappers/jumanji.py (hunk 17 above)
Comment (RuanJohn) [reply mava/wrappers/jumanji.py:61]: I also back not having all caps

File: mava/configs/env/scenario/15x15-4p-3f.yaml [hunk 22]
```diff
@@ -0,0 +1,15 @@
...
+task_config:
+  grid_size: 15
+  fov: 15
+  num_agents: 4
+  num_food: 3
+  max_agent_level: 2
+  force_coop: False
+
+# GridObserver config
+grid_obs_config:
```
[the remark this answers is not in the document]
Comment (RuanJohn) [reply mava/configs/env/scenario/15x15-4p-3f.yaml:13]: They use the vector one in the paper yes. Since we don't have any conv nets yet in mava it might even be a good idea to remove the grid obs stuff

[the remark this answers is not in the document]
Comment (WiemKhlifi) [reply mava/configs/env/scenario/15x15-4p-3f.yaml:13]: Yes the default observer is the vector one so these configs will be unnecessary since we're only considering the vector observation, I'll remove them.
